### PR TITLE
Functionality for requesting, lending and returning books - and deleting; navigation in base template

### PR DESCRIPTION
**Pull request by katjad**
Warning, mega pull request! Well, it is pretty repetitive in some way. Long mixins.py file ;) I built it on how the delete functionality works, kind of sub classed that, and I don't know how you would best organise that code, I just stuck view 'extension' in that file. - In any case it all seems to work. I realised requesting is the only action that would be done by the person borrowing, lend and return done by owner/admin.  (though disputable, but I'll better talk to you about that irl)

I had one test failing which actually alerted me to something wrong in a template! Is now fixed.

### CODE REVIEW

File: books/mixins.py [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+
+class ReturnView(SingleObjectTemplateResponseMixin, BaseLoanView):
+    """
+    View for returning a book
+    """
+
+    template_name_suffix = "_confirm_return_item"
+    success_url = "book_detail"
+
+    def createBookHolder(self, book, holder, status):
```
Comment (Geekfish) [inline books/mixins.py]: I think `createBookHolder` doesn't need to be defined anymore on `ReturnView` and `LendView`?

File: books/mixins.py [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+        self.object = self.get_object()
+        book = self.object
+        holder = self.get_holder(request, *args, **kwargs)
+        success_url = self.get_success_url(book.id)
+        status = self.get_status()
+        self.createBookHolder(book, holder, status)
+
+        return HttpResponseRedirect(success_url)
+
+    def createBookHolder(self, book, holder, status):
```
Comment (Geekfish) [inline books/mixins.py]: camel case sneaked in? ;)

File: books/mixins.py [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+        if self.allowed_status and book.status not in self.allowed_status:
+            return HttpResponseRedirect("/")
```
Comment (Geekfish) [inline books/mixins.py]: This is ok now, but we should consider dealing with failures in a more helpful way later on (raising errors, showing messages...)

File: books/mixins.py [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+        holder = self.get_holder(request, *args, **kwargs)
+        context = self.get_context_data(object=self.object, holder=holder)
+        return self.render_to_response(context)
+
+    # Add support for browsers which only accept GET and POST for now.
+    def post(self, request, *args, **kwargs):
+        return self.loan_item(request, *args, **kwargs)
+
+
+class RequestView(SingleObjectTemplateResponseMixin, BaseLoanView):
```
Comment (Geekfish) [inline books/mixins.py]: I'm not quite sure I understand why there is need for `RequestView` here, and then `BookRequest` in the book views? I would think that just `BookRequest` having these extra attributes would be ok!

Mutliple inheritence can get pretty hairy quickly, so it's usually good to avoid splitting things too much unless it really is necessary.

For example you could have this sort of setup:

```python
# everything in views.py

class BaseLoanView(SingleObjectTemplateResponseMixin, BaseDetailView):
    # Everything from LoanMixin
    # + everything already in the BaseLoanView

class RequestBook(BaseLoanView):
    template_name_suffix = "_confirm_request_item"
    success_url = "book_detail"
    model = Book
    new_status = "RQ"
    allowed_status = ["AV"]

# ...
```

You can override attributes defined much further up the inheritance stack, you don't have to subclass explicitly every ancestor in order to change it!

As a general design feedback:
It's tempting to start trying to break things into Mixins because Django provides some Mixins. But Django is trying to provide reusable functionality as a framework, an app will rarely need to be as extensible. Also many Django developers consider developing views using Mixins very hard to understand later (so when you look at a view definition you have to follow a big inheritence tree to figure out what the view really does...). So instead sometimes when you have repeated logic can simply use traditional inheritance, or turn it into a regular reusable function.

The bottom line is, don't be afraid of a bit of duplication, if it means that the code is easier to follow :)

Also, a more specific question to things as they are:
Is there a reason you choose to use `SingleObjectTemplateResponseMixin, BaseDetailView` as opposed to the more standard `DetailView`?

File: books/models.py [hunk 5]
```diff
@@ -39,6 +39,48 @@ class Book(models.Model):
...
+            latestby
+        )
+
+    def get_holder_for_status(self, status, latestby):
+        loan = self.get_loan(status, latestby)
+        return loan.holder
+
+    # Book requested
+
+    @property
```
Comment (Geekfish) [inline books/models.py]: Properties are nice syntactic sugar, but use them with care. For someone accessing an instance, a property looks exactly like an attribute. But sometimes you can write code that performs queries. So something that a developer thought was "inexpensive" (accessing a a pre-loaded object attribute), is actually "expensive" (performs one or more db lookups!).

For this reason it's better practice to have `get_xyz()` functions instead of properties for the more "expensive" calls.

A good example of a property would be to have something like:

```python
@property
def is_requested(self):
    return self.status == "RQ"
```

This performs an inexpensive evaluation.

File: templates/books/book_confirm_delete.html [hunk 6]
```diff
@@ -0,0 +1,14 @@
+{% extends 'books/base.html' %}
+
+{% block content %}
+
+{% if book.owner == user or user.is_superuser %}
+<form method="post">{% csrf_token %}
+    <p>Are you sure you want to delete "{{ object }}"?</p>
```
Comment (Geekfish) [inline templates/books/book_confirm_delete.html:7]: It's a good idea to not rely on the __str__ representation of your objects. It makes it unclear to the template (human) editor as to what is being printed.  Either edit the view `context` to include a more explicit variable (ex `book_title`), or use something like `object.title` here.

This is another disadvantage of django generic views, the name `object` is a bit vague ;)

File: books/urls.py [hunk 7]
```diff
@@ -11,6 +11,21 @@
         login_required(views.BookUpdate.as_view()),
         name="book_update",
     ),
+    path(
+        "<int:pk>/request_item/<int:userid>",
```
Comment (Geekfish) [inline books/urls.py]: You have put the "holder" of the book as part of the url here, similar to the book id.

This possibly causes some design-level concerns to be considered:

- 1 Security. The front-end tells us who the holder is. A person alters the form and can now request the book from anyone. So Mallory wants to request a book from Alice, but instead alters the form so that the userid in the url refers to his friend, Chuck. Chuck can then confirm on behalf of Alice.
- 2 Consistency. The front-end might be outdated. Bob may try to request the book from Alice, but Eve has already requested the book and Alice has already confirmed. So now the holder is actually Eve, but the outdated front-end will still request it from Alice.
- 3 "Restfull-ness". Book id refers to a resource on the server that we can request and perform actions on. User id is not really the resource being accessed, rather it's a parameter to some of the actions. When doing POST requests we generally prefer to put such parameters as part of the body, aka "post data". On a practical level, you may later want to provide additional params to these requests, which would make it inconvenient if all of them went in the url.

The main pain-points are really 1 and 2. These could be solved if you took the responsibility back to the backend, so that your view goes and figures out who the current holder is, without any input  from the front-end.

**Review by Geekfish — CHANGES_REQUESTED** [review]
Hey @katjad !

Woaah lots of cool new stuff here! ⭐️ 👍 

I added some feedback inline that maybe could have gone as general comments, but I hope it's not too much.

Things are getting more complex now, so lots of little things to consider, but looks like it's progressing well 😄

File: books/mixins.py (hunk 1 above)
Comment (katjad) [reply books/mixins.py]: They are slightly different for the different views. The current date is used for date_returned and date_borrowed respectively. But maybe that could still be packed into one function?

File: books/mixins.py (hunk 2 above)
Comment (katjad) [reply books/mixins.py]: oops. changed it! :)

File: books/mixins.py (hunk 3 above)
Comment (katjad) [reply books/mixins.py]: oh yes, that's true, I really did it as a temporary measure :)

File: books/mixins.py (hunk 4 above)
Comment (katjad) [reply books/mixins.py]: Oh yes! I had been wondering whether one step could be left out. But it's actually more than one. 

I did not understand that the functionality to have a confirm view before redirecting to the default view was already inherent in the DetailView! Of course the DetailView is already using the SingleObjectResponseMixin. I had not properly looked how it all is put together. Thanks for pointing these things out to me!! (So now I have put everything in the views.py file)

File: books/models.py [hunk 8]
```diff
@@ -39,47 +40,46 @@ class Book(models.Model):
...
+        except BookHolder.DoesNotExist:
+            loan = None
+        return loan
+
+    def get_holder_for_status(self, status):
+        loan = self.get_loan(status)
+        if loan:
+            return loan.holder
+        else:
+            return None
```
Comment (Geekfish) [inline books/models.py:58]: not a problem at all, just so you also know the "idiomatic" way to write this:
```python
loan = self.get_loan(status)
return loan and loan.holder
```

File: books/models.py [hunk 9]
```diff
@@ -39,47 +40,46 @@ class Book(models.Model):
...
+    def get_latestby_for_status(self, status):
+        status_latestby = {
+            "RQ": "date_requested",
+            "OL": "date_borrowed",
+            "AV": "date_returned",
+            "NA": None,
+        }
+        return status_latestby[status]
+
+    def get_date_for_status(self, status):
```
Comment (Geekfish) [inline books/models.py:72]: Maybe you could reuse the `status_latestby` map above (which is like "status_to_date_field" kind of), so your code here would be like:
```python
loan = self.get_loan(status)
return getattr(loan, STATUS_TO_DATE_FIELD[status], None)
```

**Review by Geekfish — APPROVED** [review]
This looks good to me , great work!

Ship! 🚢
